src/balance_reader.ts

```diff
@@ -0,0 +1,105 @@
+import BN from "bn.js";
+import Web3 from "web3";
+import {BatchExchangeInstance} from "../types/truffle-typings";
+
+const WORD_DATA_LENGTH = 64;
+
+/**
+ * Retrieves user's token balance as stored in the "balance" entry of the private exchange mapping balanceStates
+ * Value is directly read from storage relying on Solidity's layout of storage variables
+ * See https://solidity.readthedocs.io/en/develop/internals/layout_in_storage.html
+ * @param userAddress address of the user
+ * @param tokenAddress address of the token
+ * @param batchExchangeAddress address of the batch exchange
+ * @param web3Provider provider of Ethereum JavaScript API
+ * @return balance of the token for the given user as stored in balanceStates[userAddress][tokenAddress].balance
+ */
+export async function getBalanceState(
+  userAddress: string,
+  tokenAddress: string,
+  batchExchangeAddress: string,
+  web3Provider: Web3 = web3
+): Promise<BN> {
+  // TODO when Truffle depends on web3-utils version ^1.2.5:
+  // use soliditySha3Raw instead of soliditySha3 and remove type coercion
+  const BALANCESTATES_STORAGE_SLOT = "0x0";
+
+  const userBalancestatesStorageSlot = web3Provider.utils.soliditySha3(
+    {
+      type: "bytes32",
+      value: web3Provider.utils.padLeft(userAddress, WORD_DATA_LENGTH),
+    },
+    {
+      type: "bytes32",
+      value: web3Provider.utils.padLeft(
+        BALANCESTATES_STORAGE_SLOT,
+        WORD_DATA_LENGTH
+      ),
+    }
+  ) as string;
+
+  const targetStorageSlot = web3Provider.utils.soliditySha3(
+    {
+      type: "bytes32",
+      value: web3Provider.utils.padLeft(tokenAddress, WORD_DATA_LENGTH),
+    },
+    {
+      type: "bytes32",
+      value: web3Provider.utils.padLeft(
+        userBalancestatesStorageSlot,
+        WORD_DATA_LENGTH
+      ),
+    }
+  ) as string;
+
+  const storageAtSlot = await web3Provider.eth.getStorageAt(
+    batchExchangeAddress,
+    targetStorageSlot
+  );
+  return web3Provider.utils.toBN(storageAtSlot);
+}
+
+/**
+ * Computes amount of a token that a user can immediately withdraw from the exchange
+ * It not only checks whether a withdrawal is pending, but also considers the balance available to the user,
+ * pending deposits, and whether there were recent trades that would make withdrawing fail.
+ * @param userAddress address of the user
+ * @param tokenAddress address of the token
+ * @param batchExchange object representing the batch exchange smart contract
+ * @param web3Provider provider of Ethereum JavaScript API
+ * @return amount of token that the user would receive by calling batchExchange.withdraw
+ */
+export async function getWithdrawableAmount(
+  userAddress: string,
+  tokenAddress: string,
+  batchExchange: BatchExchangeInstance,
+  web3Provider: Web3 = web3
+): Promise<BN> {
+  const [
+    balanceState,
+    pendingDeposit,
+    pendingWithdrawal,
+    lastCreditBatchId,
+    batchId,
+  ] = await Promise.all([
```

Comment:
nit: I think even faster than parallel promises would be batching the calls into a single RPC call via [batch requests](https://github.com/ethereum/wiki/wiki/JavaScript-API#batch-requests). Not necessary in this PR but could be something cool to add to your toolbox and we should probably use this in more parts of our codebase.

Reply:
I agree, batch requests look handy. I'm not going to do it now however, the type we use for Web3 does not appear to support it (`Property 'createBatch' does not exist on type 'Web3'`). I'll keep that in mind if we'll ever  make a run of request optimizations in the future.